util/cloud/s3.cc

```diff
@@ -249,7 +249,13 @@ io::Result<io::ReadonlyFile*> S3Bucket::OpenReadFile(string_view path,
   } else {
     full_path = absl::StrCat(bucket_, "/", path);
   }
-  return OpenS3ReadFile(region_, full_path, aws_, http_client_.get(), opts);
+
+  unique_ptr http_client = std::move(http_client_);
+  error_code ec = Connect(http_client->connect_timeout_ms());
+  if (ec)
+    return make_unexpected(ec);
+
+  return OpenS3ReadFile(region_, full_path, aws_, http_client.release(), opts);
```

Comment:
Can we modify it so `OpenS3ReadFile` receives a `unique_ptr` instead of a raw pointer? This might seem innocent, but this is problematic because when you lose exception safety until you construct the new `unique_ptr` in `S3WriteFile::S3WriteFile`: If an exception happens somewhere between this pointer will be leaked.

Reply:
Yep sure - will update